proposals/extend_foreach_to_cover_generate_rules.md

```diff
@@ -0,0 +1,302 @@
+
+# Proposal
+
+Based on the use cases mentioned above, we can introduce `foreach` for generate rules in two phases:
+- Phase 1: support the generation or cloning of a list of resources for each generation rule by adding `dataList` or `cloneList` to the generate pattern; 
+- Phase 2: support iteration through the given list using the `foreach` declaration.
+
+Let's break down each phase to following:
+
+## Phase 1 - support `dataList` and `cloneList`
+
+To allow generation of multiple downstream resources for a single trigger, `dataList` and `cloneList` can be added to `generate` pattern.
+
+The following rule snippet showcases generation of a `NetworkPolicy` and a `ConfigMap` using `dataList`:
+```yaml
+spec:
+  rules:
+  - name: generate-multiple-targets
+    match:
+      any:
+      - resources:
+          kinds:
+          - Namespace
+    generate:
+      dataList:
+      - apiVersion: v1
+        kind: NetworkPolicy
+        name: default-networkpolicy
+        namespace: "{{ request.object.metadata.name }}"
+        data:
+          spec:
+            podSelector: {}
+            policyTypes:
+            - Ingress
+            - Egress
+      - apiVersion: v1
+        kind: ConfigMap
+        name: default-configmap
+        namespace: "{{ request.object.metadata.name }}"
+        data:
+          foo: bar
+```
+
+The following rule clones a source secret `regcred` and a configmap `default-configmap` from `default` namespace to the new created namespace using `cloneList`. Note that the [cloneList](https://github.com/kyverno/kyverno/blob/main/api/kyverno/v1/common_types.go#L736) is already supported which allows cloning sources by label selectors. To make it backward compatible, this proposal introduces new `list` under `cloneList` pattern to provide flexibility when selecting sources.
+
+```yaml
+spec:
+  rules:
+  - name: clone-multiple-sources
+    match:
+      any:
+      - resources:
+          kinds:
+          - Namespace
+    generate:
+      cloneList:
+        list:
+        - source:
+            namespace: default
+            name: default-regcred
+            kind: v1
+            apiVersion: Secret
+          target:
+            namespace: "{{ request.object.metadata.name }}"
+            name: regcred
+        - source:
+            namespace: default
+            name: default-configmap
+            kind: v1
+            apiVersion: ConfigMap
+          target:
+            namespace: "{{ request.object.metadata.name }}"
+            name: cloned-configmap
+```
+
+## Phase 2 - support `foreach` declaration
+
+In the second phase, the generate rule can be expanded to support `foreach` declaration in order to iterate through given list. This could be done by adding the `foreach` key under generate pattern, and a `list` attribute which is a JMESPath expression that defines the sub-elements it processes. For every `foreach` pattern, either `dataList` or `cloneList` will be supported.
+
```

Comment:
Would `data` and `clone` also be supported under the `foreach`?

Ideally all behaviors / constructs in the regular generate are also supported when using foreach.

Reply:
>Would data and clone also be supported under the foreach?

No, as the same functionalities to generate or clone single resources can be achieved using `dataList` or `cloneList`. I'm thinking whether we want to deprecate `data` and `clone` once we introduce new fields, but that requires migration of existing policies to use new structure. They are good things to discuss but it's out of the scope of this KDP. I can start a separate design discussion once we add `dataList` and `cloneList`, what do you think?

Reply:
I am fine with deprecating `data` and `clone`.

However, we should try and reuse the generate logic completely. The `foreach` is just a loop that invokes generate multiple times.

Hence, may be easier to retain support for  `data` and `clone` until they are removed.